compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -2609,7 +2609,7 @@ public Boolean visit(BRecordType t, BType s) {
             for (BField sourceField : source.fields.values()) {
                 if (t.fields.containsKey(sourceField.name.value)) {
                     BField targetField = t.fields.get(sourceField.name.value);
-                    if (isSameType(sourceField.type, targetField.type, this.unresolvedTypes) &&
+                    if (isSameType(sourceField.type, targetField.type, new HashSet<>(this.unresolvedTypes)) &&
```

Comment:
I think this algorithm is logically wrong. The unresolved list can have pairs which are totally different but due to `unresolvedTypes`  cache we say they are the same if found again in the cache. int, int|string|boolean is such a pair in an unresolved list during the original issue. This cache ideally should be utilized to work when there is a cyclic type.
What if we remove pairs from cache if they are not matching. This will also be a slightly better improvement not a good solution. 
eg :  We can change `isSameType` to remove bad pairs in the end.
```Java
        TypePair pair = new TypePair(source, target);
        if (unresolvedTypes.contains(pair)) {
            return true;                                        // THIS IS TRUE FOR CYCLIC TYPES, NOT FOR OTHER TYPES. 
        }
        unresolvedTypes.add(pair);
        BTypeVisitor<BType, Boolean> sameTypeVisitor = new BSameTypeVisitor(unresolvedTypes);
        boolean result = target.accept(sameTypeVisitor, source);
        if (result) {
            return true;
        }
        unresolvedTypes.remove(pair);
        return false;
```

Reply:
I agree to look at this complexity later. I will create an issue to track this.

Reply:
thanks!.

Reply:
New issue https://github.com/ballerina-platform/ballerina-lang/issues/36069